code/go/internal/validator/folder_item_spec.go

```diff
@@ -0,0 +1,166 @@
+package validator
+
+import (
+	"encoding/json"
+	"fmt"
+	"gopkg.in/yaml.v3"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"regexp"
+
+	"github.com/pkg/errors"
+	"github.com/xeipuuv/gojsonschema"
+)
+
+type folderItemSpec struct {
+	Description      string `yaml:"description"`
+	ItemType         string `yaml:"type"`
+	ContentMediaType string `yaml:"contentMediaType"`
+	Name             string `yaml:"name"`
+	Pattern          string `yaml:"pattern"`
+	Required         bool   `yaml:"required"`
+	Ref              string `yaml:"$ref"`
+	Visibility       string `yaml:"visibility" default:"public"`
+	commonSpec       `yaml:",inline"`
+}
+
+type itemSchemaSpec struct {
+	Spec map[string]interface{} `json:"spec" yaml:"spec"`
+}
+
+func (s *folderItemSpec) matchingFileExists(files []os.FileInfo) (bool, error) {
+	if s.Name != "" {
+		for _, file := range files {
+			if file.Name() == s.Name {
+				return s.isSameType(file), nil
+			}
+		}
+	} else if s.Pattern != "" {
+		for _, file := range files {
+			isMatch, err := regexp.MatchString(s.Pattern, file.Name())
+			if err != nil {
+				return false, errors.Wrap(err, "invalid folder item spec pattern")
+			}
+			if isMatch {
+				return s.isSameType(file), nil
+			}
+		}
+	}
+
+	return false, nil
+}
+
+func (s *folderItemSpec) isSameType(file os.FileInfo) bool {
+	switch s.ItemType {
+	case itemTypeFile:
+		return !file.IsDir()
+	case itemTypeFolder:
+		return file.IsDir()
+	}
+
+	return false
+}
+
+func (s *folderItemSpec) validate(fs http.FileSystem, folderSpecPath string, itemPath string) ValidationErrors {
+	if s.Ref == "" {
+		return nil // no item's schema defined
```

Comment:
What is the schema is defined inline and not referenced via `$ref`?

Reply:
not sure if I follow you...
this might be bad wording, should be: schema reference not provided. Does it sound better?

Reply:
Sorry, I was referring to needing something like this: https://github.com/elastic/package-spec/blob/master/code/go/internal/validator/folder.go#L125-L141. Maybe this is not the right place for it but then it should be added where this method is called from.

Reply:
Ah, now I see... I totally missed the case where schema is defined inline, didn't know about this. I will improve the implementation.

Reply:
@ycombinator Hm.. I dived into this topic and got a bit confused. Do we really need it? What would a use case for this? This `$ref` refers to a place in folder schema in which we refer to item's schema. It's not part of the JSON-schema, so I believe we can enforce to keep it in a separate file (to separate concerns and prevent from creating a huge all-in-one files).

tl;dr `$ref` in folder schema is always external.

WDYT?

Reply:
So in the folder schema, there can be two `type`s of items: `folder` (for describing the structure of a sub-folder within the folder) and `file` (for describing the structure of a file within the folder). 

For `type: folder`, we already have a couple of places in the spec where use `$ref` ([example](https://github.com/elastic/package-spec/blob/f45b05c887b5d62dea8f5c21577520a0e113716c/versions/1/dataset/spec.yml#L20)) and a couple other places where we inline the definition instead of using `$ref` ([example](https://github.com/elastic/package-spec/blob/f45b05c887b5d62dea8f5c21577520a0e113716c/versions/1/dataset/spec.yml#L26)).

For `type: file`, I checked, and _at the moment_ we only have places where we use `$ref` (example). We do not yet have any places where we are inlining the spec (aka schema) for the file itself. However, I can see this being useful for simple file specs ([example](https://github.com/elastic/package-spec/blob/master/versions/1/dataset/_dev/test/pipeline/event.spec.yml)).

 So I think there is a use case for it and we should support it.

Also, if we allowing inline specs for sub-folder items but not for file items, we are introducing an inconsistency. We could, of course, resolve this inconsistency by going the other way: that is, we can say that we _only_ allow `$ref`s for item specs, regardless of whether those are folder items or file items; that we do not accept inline specs for either item type. I think this is slightly inconvenient when the item has a small spec but I'd be okay with this approach too. If you prefer this, then let's make a separate PR to first remove support for inline specs for `type: folder` items and adjust the spec files accordingly.